src/qibolab/instruments/qm.py

```diff
@@ -0,0 +1,598 @@
+                # register readout pulse in elements
+                self.config["elements"][f"readout{qubit}"]["operations"][pulse.serial] = pulse.serial
+
+            else:
+                raise_error(TypeError, f"Unknown pulse type {pulse.type.name}.")
+
+        return f"{pulse.type.name.lower()}{str(qubit)}"
+
+    def register_waveform(self, pulse, mode="i"):
+        """Registers waveforms in QM config.
+
+        QM supports two kinds of waveforms, examples:
+            "zero_wf": {"type": "constant", "sample": 0.0}
+            "x90_wf": {"type": "arbitrary", "samples": x90_wf.tolist()}
+
+        Args:
+            pulse (:class:`qibolab.pulses.Pulse`): Pulse object to read the waveform from.
+            mode (str): "i" or "q" specifying which channel the waveform will be played.
+
+        Returns:
+            serial (str): String with a serialization of the waveform.
+                Used as key to identify the waveform in the config.
+        """
+        from qibolab.pulses import Rectangular
+
+        waveforms = self.config["waveforms"]
+        # Maybe need to force zero q waveforms
+        # if pulse.type.name == "READOUT" and mode == "q":
+        #    serial = "zero_wf"
+        #    if serial not in waveforms:
+        #        waveforms[serial] = {"type": "constant", "sample": 0.0}
+        if isinstance(pulse.shape, Rectangular):
+            serial = f"constant_wf{pulse.amplitude}"
+            if serial not in waveforms:
+                waveforms[serial] = {"type": "constant", "sample": pulse.amplitude}
+        else:
+            waveform = getattr(pulse, f"envelope_waveform_{mode}")
+            serial = waveform.serial
+            if serial not in waveforms:
+                waveforms[serial] = {"type": "arbitrary", "samples": waveform.data.tolist()}
+        return serial
+
+    def register_integration_weights(self, qubit, readout_len):
+        """Registers integration weights in QM config.
+
+        Args:
+            qubit (:class:`qibolab.platforms.quantum_machines.Qubit`): Qubit
+                object that the integration weights will be used for.
+            readout_len (int): Duration of the readout pulse in ns.
+        """
+        rotation_angle = qubit.characterization.rotation_angle
+        self.config["integration_weights"].update(
+            {
+                f"cosine_weights{qubit}": {
+                    "cosine": [(np.cos(rotation_angle), readout_len)],
+                    "sine": [(-np.sin(rotation_angle), readout_len)],
+                },
+                f"sine_weights{qubit}": {
+                    "cosine": [(np.sin(rotation_angle), readout_len)],
+                    "sine": [(np.cos(rotation_angle), readout_len)],
+                },
+                f"minus_sine_weights{qubit}": {
+                    "cosine": [(-np.sin(rotation_angle), readout_len)],
+                    "sine": [(-np.cos(rotation_angle), readout_len)],
+                },
+            }
+        )
```

Comment:
It would be much more natural if they shared the same type, and they were collected in a single `qubit` container, instead of containing the indication in the string:
```py
Weight = List[Tuple[float, int]]

@dataclass
class Weights:
    cosine: Weight
    sine: Weight
```

```suggestion    
        self.config["integration_weights"][qubit] = {
            "cosine": Weights(cosine=[(np.cos(rotation_angle), readout_len)], sine=[(-np.sin(rotation_angle), readout_len)]),
            "sine": ...,
            "minus_sine": ...
        }
```

Reply:
Moreover, since the information is always the same (i.e. `rotation_angle` and `readout_len`), it would be even better if you turn them in properties, instead of entries (reusing all the same info):

```py
Weight = List[Tuple[float, int]]

@dataclass
class Weights:
    cosine: Weight
    sine: Weight

@dataclass
class QubitWeights: 
    angle: float
    readout_len: int
    
    @property
    def cosine(self):
        return Weights(cosine=[(np.cos(self.angle), self.readout_len)], sine=[(-np.sin(self.angle), self.readout_len)])
        
    @property
    def sine(self):
        ...
        
    @property
    def minus_sign(self):
        ...
```

```suggestion
        self.config["integration_weights"][qubit] = QubitWeights(angle=rotation_angle, readout_len=readout_len)
```

Reply:
Of course class name are always placeholders, please most suitable (or your favorite) ones.